### PR TITLE
[RFC] stm32h7: Enable support for optional/alternate clocks

### LINKED ISSUE #41650: STM32H7 SPI123 incorrect clock source used for prescaler calculation
## Describe the bug
In the STM32H7 series the assumed clock source is incorrect when using SPI1, 2 or 3. This incorrect clock source is then used to calculate the clock prescaler (`SPI_CFG:MBR`), which can then potentially be calculated incorrectly. More specifically, it assumes the SCK clock source is, for example `APB2`, while it's by actually by default `PLL1Q`. This resulted that our custom STM32H743 implementation is outputting an incorrect SPI clock frequency.

This is because for the H7 series the SPI1, 2, and 3 has a _seperate_ clock domain that governs the SCK speed. See the following snippet from RM0433 Figure 609. 

image

Here `spi_ker_ck` is the `SCLK` source, not `PCLK` (f.i. `APB2`). The value of the `spi_ker_ck` is set by `RCC_D2CCIP1R : SPI123SEL`, which is `pll1_q_ck` by default.

## To Reproduce
We can slightly adjust the `nucleo_h743zi` DTS and abuse the `led_ws2812` app to easily see this behaviour.
1. I'm running this commit from the main branch `e2588d6a42`
2. Slightly adjust your code by either:
  a. Checking out [this branch](https://github.com/fancom/zephyr/tree/spi123-incorrect-clock)
  b. Applying the patch `nucleo-configuration-patch` supplied below.
    - [Optional] Add `print-clocks-output` patch to print out debugging clock information
4. `west build -p auto -b nucleo_h743zi samples/drivers/led_ws2812/`
5. `west flash` onto a `nucleo_h743zi`
6. Probe `CN7:10` (`D13` or `SPI_A_SCK` or `SPI_1_SCK`) with an oscilloscope
7. Witness the `SCK` frequency incorrectly being 6MHz instead of the desired and calculated 3MHz.

It will also output the following with the optional logging patch 
```
[00:00:00.005,000] <err> spi_ll_stm32: Is this SPI1? 1
[00:00:00.005,000] <err> spi_ll_stm32: spi clock bus: STM32_CLOCK_BUS_APB2
[00:00:00.005,000] <err> spi_ll_stm32: STM32_CLOCK_BUS_APB2 freq: 96000000
[00:00:00.005,000] <err> spi_ll_stm32: LL_RCC_SPI123_CLKSOURCE: LL_RCC_SPI123_CLKSOURCE_PLL1Q
[00:00:00.005,000] <err> spi_ll_stm32: pllq freq: 192000000
[00:00:00.005,000] <err> spi_ll_stm32: Desired frequency: 3000000
[00:00:00.005,000] <err> spi_ll_stm32: BR: 5 should create clock of freq: 3000000
[00:00:00.005,000] <err> spi_ll_stm32: Resulting actual clock clock with PLL: 6000000
````

## Expected behavior
Expect the prescaler calculation to use the correct clock source to output the desired frequency.

## Impact
Hard to notice issues with incorrect SPI frequencies meaning devices run slower/faster than designed for.

## Environment (please complete the following information):
 - OS: Linux
 - Toolchain: Zephyr SDK 0.13.2
 - Commit SHA: `e2588d6a42`

## Additional Context
I created #40998 to attempt to fix this issue, but ABOSTM found a showstopping flaw in my implementation. We then proposed a new sollution would be to:

- Use the default `spi_ker_ck` value, since it's not possible to set two clocks the DTS. This means SCK source will always be `PLL1Q`.
- When calculating the prescaler use PLL1Q clock value.
- Add checks to ensure `PPL1Q` is enabled and `#error` if not.

I will try to create a fix for this implementation soon. However, I'm not sure yet what would be the best way to calculate the `PLL1Q` frequency, and would like some thoughts on that. The way I did it in the patch below is quite ugly.

It's coincidance that this has gone unnoticed during previous SPI driver testing. If I understand correctly then the Nucleo H743zi is popular for STM32H7 testing, and the `PLL1 Q` divider usually sets the clock up by chance that `PLL1Q == APB2`. Therefore I had to change the dts in the patch below to change `pll1_q` divisor from `4` to `1` to highlight this issue.

### **nucleo-configuration-patch**
```
From 21c48a1f7ec70e773cf7639c3d0a3175ab390ff3 Mon Sep 17 00:00:00 2001
From: Hein Wessels <hein.wessels@nobleo.nl>
Date: Fri, 7 Jan 2022 14:19:30 +0100
Subject: [PATCH] nucleo config

---
 .../led_ws2812/boards/nucleo_h743.conf        |  2 +
 .../led_ws2812/boards/nucleo_h743zi.overlay   | 39 +++++++++++++++++++
 2 files changed, 41 insertions(+)
 create mode 100644 samples/drivers/led_ws2812/boards/nucleo_h743.conf
 create mode 100644 samples/drivers/led_ws2812/boards/nucleo_h743zi.overlay

diff --git a/samples/drivers/led_ws2812/boards/nucleo_h743.conf b/samples/drivers/led_ws2812/boards/nucleo_h743.conf
new file mode 100644
index 0000000000..319a42ce98
--- /dev/null
+++ b/samples/drivers/led_ws2812/boards/nucleo_h743.conf
@@ -0,0 +1,2 @@
+CONFIG_WS2812_STRIP_SPI=y
+CONFIG_SPI=y
diff --git a/samples/drivers/led_ws2812/boards/nucleo_h743zi.overlay b/samples/drivers/led_ws2812/boards/nucleo_h743zi.overlay
new file mode 100644
index 0000000000..0d1fba0ac4
--- /dev/null
+++ b/samples/drivers/led_ws2812/boards/nucleo_h743zi.overlay
@@ -0,0 +1,39 @@
+/*
+ * Copyright (c) 2021, STMicroelectronics
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <dt-bindings/led/led.h>
+
+arduino_spi: &spi1 {};
+
+
+&arduino_spi {
+	led_strip: ws2812@0 {
+		compatible = "worldsemi,ws2812-spi";
+		label = "WS2812";
+
+		/* SPI */
+		reg = <0>; /* ignored, but necessary for SPI bindings */
+		spi-max-frequency = <3000000>;
+
+		/* WS2812 */
+		chain-length = <16>; /* arbitrary; change at will */
+		spi-one-frame = <0x70>;
+		spi-zero-frame = <0x40>;
+		color-mapping = <LED_COLOR_ID_GREEN
+				 LED_COLOR_ID_RED
+				 LED_COLOR_ID_BLUE>;
+	};
+};
+
+/ {
+	aliases {
+		led-strip = &led_strip;
+	};
+};
+
+&pll{
+	div-q = <1>;
+};
\ No newline at end of file
-- 
2.25.1

```
### **print-clocks-output**

```
From e376b740607da0f007265748cdae4defbb02bc1e Mon Sep 17 00:00:00 2001
From: Hein Wessels <hein.wessels@nobleo.nl>
Date: Fri, 7 Jan 2022 14:46:26 +0100
Subject: [PATCH] Add log output

---
 drivers/spi/spi_ll_stm32.c | 94 ++++++++++++++++++++++++++++++++++++++
 1 file changed, 94 insertions(+)

diff --git a/drivers/spi/spi_ll_stm32.c b/drivers/spi/spi_ll_stm32.c
index 47a5ae01d6..d779329ca0 100644
--- a/drivers/spi/spi_ll_stm32.c
+++ b/drivers/spi/spi_ll_stm32.c
@@ -451,6 +451,72 @@ static void spi_stm32_isr(const struct device *dev)
 }
 #endif
 
+#include <stm32_ll_rcc.h>
+
+#define PLLQ_FREQ(pllsrc_freq, divm, divn, divq)	(((pllsrc_freq)*\
+							(divn))/((divm)*(divq)))
+
+static inline uint32_t get_pllsrc_frequency(void)
+{
+	switch (LL_RCC_PLL_GetSource()) {
+	case LL_RCC_PLLSOURCE_HSI:
+		return HSI_VALUE;
+	case LL_RCC_PLLSOURCE_CSI:
+		return CSI_VALUE;
+	case LL_RCC_PLLSOURCE_HSE:
+		return HSE_VALUE;
+	case LL_RCC_PLLSOURCE_NONE:
+	default:
+		return 0;
+	}
+}
+
+static uint32_t get_pllq_frequency(void)
+{
+	uint32_t sysclk = 0;
+	uint32_t hpre = 0;
+	uint32_t hsidiv = 0;
+
+	/* Get the current HSI divider */
+	switch (LL_RCC_HSI_GetDivider()) {
+	case LL_RCC_HSI_DIV2:
+		hsidiv = 2;
+		break;
+	case LL_RCC_HSI_DIV4:
+		hsidiv = 4;
+		break;
+	case LL_RCC_HSI_DIV8:
+		hsidiv = 8;
+		break;
+	case LL_RCC_HSI_DIV1:
+	default:
+		hsidiv = 1;
+		break;
+	}
+
+	/* Get the current system clock source */
+	switch (LL_RCC_GetSysClkSource()) {
+	case LL_RCC_SYS_CLKSOURCE_STATUS_HSI:
+		sysclk = HSI_VALUE/hsidiv;
+		break;
+	case LL_RCC_SYS_CLKSOURCE_STATUS_CSI:
+		sysclk = CSI_VALUE;
+		break;
+	case LL_RCC_SYS_CLKSOURCE_STATUS_HSE:
+		sysclk = HSE_VALUE;
+		break;
+	case LL_RCC_SYS_CLKSOURCE_STATUS_PLL1:
+		sysclk = PLLQ_FREQ(get_pllsrc_frequency(),
+				   LL_RCC_PLL1_GetM(),
+				   LL_RCC_PLL1_GetN(),
+				   LL_RCC_PLL1_GetQ());
+		break;
+	}
+
+	return sysclk;
+
+}
+
 static int spi_stm32_configure(const struct device *dev,
 			       const struct spi_config *config)
 {
@@ -486,13 +552,41 @@ static int spi_stm32_configure(const struct device *dev,
 		return -EIO;
 	}
 
+	LOG_ERR("Is this SPI1? %d", spi == SPI1);
+	LOG_ERR("spi clock bus: %s", cfg->pclken.bus == 3 ? "STM32_CLOCK_BUS_APB2" : "something else" );
+	LOG_ERR("STM32_CLOCK_BUS_APB2 freq: %d", clock);
+
+	switch(LL_RCC_GetSPIClockSource(LL_RCC_SPI123_CLKSOURCE)){
+		case LL_RCC_SPI123_CLKSOURCE_PLL1Q:
+			LOG_ERR("LL_RCC_SPI123_CLKSOURCE: %s", "LL_RCC_SPI123_CLKSOURCE_PLL1Q");
+			break;
+		case LL_RCC_SPI123_CLKSOURCE_PLL2P:
+			LOG_ERR("LL_RCC_SPI123_CLKSOURCE: %s", "LL_RCC_SPI123_CLKSOURCE_PLL2P");
+			break;
+		case LL_RCC_SPI123_CLKSOURCE_PLL3P:
+			LOG_ERR("LL_RCC_SPI123_CLKSOURCE: %s", "LL_RCC_SPI123_CLKSOURCE_PLL3P");
+			break;
+		case LL_RCC_SPI123_CLKSOURCE_I2S_CKIN:
+			LOG_ERR("LL_RCC_SPI123_CLKSOURCE: %s", "LL_RCC_SPI123_CLKSOURCE_I2S_CKIN");
+			break;
+		case LL_RCC_SPI123_CLKSOURCE_CLKP:
+			LOG_ERR("LL_RCC_SPI123_CLKSOURCE: %s", "LL_RCC_SPI123_CLKSOURCE_CLKP");
+			break;			
+	}	
+	LOG_ERR("pllq freq: %d", get_pllq_frequency());
+
+	LOG_ERR("Desired frequency: %d", config->frequency);
+
 	for (br = 1 ; br <= ARRAY_SIZE(scaler) ; ++br) {
 		uint32_t clk = clock >> br;
 
 		if (clk <= config->frequency) {
+			LOG_ERR("BR: %d should create clock of freq: %d", br, clk);
 			break;
 		}
 	}
+	
+	LOG_ERR("Resulting actual clock clock with PLL: %d", get_pllq_frequency() >> br);	
 
 	if (br > ARRAY_SIZE(scaler)) {
 		LOG_ERR("Unsupported frequency %uHz, max %uHz, min %uHz",
-- 
2.25.1
```

### PR DESCRIPTION
**Pull request by erwango**
## TL;DR ##
STM32 clock_control driver update (limited to H7 for now) to provide a way to configure alternate clocks on peripherals that support it:
```
    clocks = <&rcc STM32_CLOCK_BUS_APB2 0x00001000>,  <- Existing, bus driven, "register" clock
             <&rcc STM32_SRC_PLL3_P SPI123_SEL(2)>;   <- Newly added alternate clock (PLL3_P output here)
```
This alternate clock source could be selected in peripheral driver thanks to new `clock_control_configure()` API:
```
	if (IS_ENABLED(STM32_SPI_OPT_CLOCK_SUPPORT) && (cfg->pclk_len > 1)) {
		if (clock_control_configure(DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE),
					    (clock_control_subsys_t) &cfg->pclken[1],
					    NULL) != 0) {
			LOG_ERR("Could not select alternate SPI source clock");
			return -EIO;
		}
	}
```

## Goal
Aim of this work is to enable the use of an optional clock in a device configuration.
This is particularly required in case of STM32H7 based platform on which device clock configuration is usually based on 2 different clocks (cf #41650)
- A register clock, required for device configuration, usually coming from APB/AHB bus.
- A kernel clock, which will be used to clock the hardware block (and hence is the one driving the device frequency), typically derived from an alternate clck source (PLL output, fixed clock, ..)

Fixes #41650

### Why it worked so far on H7
Device clock configuration used to be functional up to now since clock configuration was set so that APB clk and alternate device clk (the default reset one) were equivalent. 

### Secondary goal
Besides the initial goal, idea was to provide the basis to support equivalent functionalities on onther STM32 series, such as the possibility to provide an alternate clock to a peripheral. For instance, configuring LSE as the LPUART or LPTIM clock, or HSI48 as the USB clock.

## Update of clock_control API (03/14):
 
Reworked to make use of new `clock_control_configure()` API, cf https://github.com/zephyrproject-rtos/zephyr/pull/43790

### Alternate clocks bindings
Additional/Alternate clocks could be fixed clocks (HSI, HSE, ..), PLL outputs (PLLX_P/Q/R), sysclock, ....
For all these clocks, I've decided to keep existing RCC device driver as the clock controller. As a consequence, the dt binding used to select a specific clock looks to:
```
	clocks = <&rcc STM32_CLOCK_BUS_APB2 0x00001000>,  <- Existing, bus driven, "register" clock
		 <&rcc STM32_SRC_PLL3_P SPI123_SEL(2)>;   <- Newly added alternate clock (PLL3_P output here)
```
Another way to support this would have been to set up each potential clock source as a clock controler on its own and use the following binding:
```
	clocks = <&rcc STM32_CLOCK_BUS_APB2 0x00001000>,  <- Existing, bus driven, "register" clock
		 <&pll3 pll_p_output SPI123_SEL(2)>;      <- Discarded binding option
```
I've discarded this second option because, to be consistent with the binding, this second option would require to instantiate each potential source clock as a device driver, each one implementing clock_control API. This would have had negative impact on code footprint, which could have been considered ok for STM32H7 series but problematic on "smaller" STM32 series.

### Clock mux bindings
Complete solution requires to find a way to configure clock mux (as per_ck).

First option is to add a dedicated node descibing  the clock mux. Its only property is to configure its clock input
```
&perck {
	clocks = <&rcc STM32_SRC_HSI_KER CKPER_SEL(0)>;
	status = "okay";
};
```
Works fine, main drawback is the cost (~220 bytes of flash)

A much simpler alternative was to add clock configuration directly to the clock consumer end node:
```
	clocks = <&rcc STM32_CLOCK_BUS_APB2 0x00001000>,
		 <&rcc STM32_SRC_HSI_KER CKPER_SEL(0)>,
		 <&rcc STM32_SRC_CKPER SPI123_SEL(4)>;
```
Though, this proposal has 2 main drawbacks:
- If board dts is poorly configured, 2 consumers of the same internal clock (per_ck here) could set conflicting configurations
- This adds complexity on the consumers driver implementation side that should be able to differentiate between all these clocks 

In the end the addition of a clock mux node device felt saner.

### Dependencies
This work is based on:
- ~~https://github.com/zephyrproject-rtos/zephyr/pull/41585, which contain the preliminary clock_control driver work~~ merged
- ~~New set of `/include/devicetree/clock.h` macros (cf https://github.com/zephyrproject-rtos/zephyr/pull/42326)~~ merged
- Dedicated tests provided in `tests/drivers/clock_control/stm32_clock_configuration/stm32h7_devices/`

### Limitations and next steps:
- Only STM32 SPI driver is made aware of this. All STM32 drivers should be updated to take this new clock config into account.
- Extend to other STM32 series. This kernel clock is generally not available on other STM32 series, but most of this work can be reused to support the configuration of alternate peripheral clocks (like HSI on LPUART). This extension would first require a clock_control driver clean up similar to #41585.
- It is not possible today to `clock_off` supplementary clocks as they potentially clocks multiple peripherals (or even the core itself). So, this most likely requires the use of power-domains

### CODE REVIEW

**Comment by erwango** [comment]
Added clock mux support following discussion on L5 case [here](https://github.com/zephyrproject-rtos/zephyr/issues/42374)

File: drivers/clock_control/clock_stm32_ll_h7.c [hunk 1]
```diff
@@ -313,97 +313,43 @@ static inline int stm32_clock_control_on(const struct device *dev,
...
+	reg = (uint32_t *)(DT_REG_ADDR(DT_NODELABEL(rcc)) + pclken->bus);
+	reg_val = *reg;
+	reg_val &= ~pclken->enr;
+	reg_val |= pclken->enr;
+	*reg = reg_val;
 
 	z_stm32_hsem_unlock(CFG_HW_RCC_SEMID);
 
-	return rc;
+	return 0;
```
Comment (FRASTM) [inline drivers/clock_control/clock_stm32_ll_h7.c:431]: is there any benefit to raise an error code here in case the DTS is malformed ?

File: tests/drivers/clock_control/stm32_clock_configuration/stm32h7_devices/boards/spi1_pll3p_1_d1ppre_4.overlay [hunk 2]
```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright (c) 2021 Linaro Limited
+ *
```
Comment (FRASTM) [inline tests/drivers/clock_control/stm32_clock_configuration/stm32h7_devices/boards/spi1_pll3p_1_d1ppre_4.overlay:3]: * Copyright (c) 202**2**

**Review by FRASTM — DISMISSED** [review]

File: drivers/clock_control/clock_stm32_ll_h7.c (hunk 1 above)
Comment (erwango) [reply drivers/clock_control/clock_stm32_ll_h7.c:431]: In a next commit, a check is added to verify that the requested clock is enabled.
Regarding the other parameters I'm not sure we can do much more checks.

File: include/drivers/clock_control.h [hunk 3]
```diff
@@ -150,6 +151,39 @@ static inline int clock_control_off(const struct device *dev,
...
+ *
+ * It is client's reposnsibility to ensure that subsequent calls to the API
+ * provides the right information to enable clock_control driver to perform
+ * the right action (such as using the right clock source on _get_rate call).
+ *
+ * @param dev Device structure whose driver controls the clock
+ * @param sys Opaque data representing the clock
+ * @return 0 on success, negative errno on failure.
+ */
+static inline int clock_control_select(const struct device *dev,
```
Comment (erwango) [inline include/drivers/clock_control.h:168]: @nordic-krch Would you mind commenting this approval ?
I've added it on top of current PR, but I plan to raise it separately once we're aligned (and then rebase current PR on top)

**Comment by erwango** [comment]
Rework to use new `clock_control_configure` API

**Comment by erwango** [comment]
@galak, @gmarull I'd welcome your inputs if you have time.
